### PR TITLE
Pipeline store writes

### PR DESCRIPTION
**Pull request by lutter**
This PR makes it possible to perform database writes in parallel with the rest of processing during indexing. The size of the write queue can be set with the environment variable `GRAPH_STORE_WRITE_QUEUE` which defaults to 5 write or revert operations. Setting this to 0 makes writing synchronously and reinstates the previous behavior, also bypassing a bunch of new code.

### CODE REVIEW

**Comment by leoyvens** [comment]
We had agreed to do point 2 of this comment https://github.com/graphprotocol/graph-node/pull/3084#issuecomment-999017779 where the instance manager keeps track of its subgraph pointer rather than querying the writable agent for it. Does that still sound like a good idea to you?

**Comment by lutter** [comment]
> We had agreed to do point 2 of [#3084 (comment)](https://github.com/graphprotocol/graph-node/pull/3084#issuecomment-999017779) where the instance manager keeps track of its subgraph pointer rather than querying the writable agent for it. Does that still sound like a good idea to you?

It does, I just didn't want to hold up this PR by adding that in, too. It'll require returning subgraph pointer/cursor from `start_subgraph_deployment` and needs some changes in the instance manager to thread that through IIRC.

**Review by leoyvens — CHANGES_REQUESTED** [review]
Big brain PR 🧠

File: graph/src/components/store.rs [hunk 1]
```diff
@@ -1105,6 +1108,9 @@ pub trait WritableStore: Send + Sync + 'static {
     async fn health(&self, id: &DeploymentHash) -> Result<SubgraphHealth, StoreError>;
 
     fn input_schema(&self) -> Arc<Schema>;
+
+    /// Wait for the background writer to finish processing its queue
+    async fn wait(&self) -> Result<(), StoreError>;
```
[the remark this answers is not in the document]
Comment (lutter) [reply graph/src/components/store.rs:1113]: Yep, changed that.

File: store/postgres/src/subgraph_store.rs [hunk 2]
```diff
@@ -1077,7 +1088,12 @@ impl SubgraphStoreTrait for SubgraphStore {
         id: &DeploymentHash,
     ) -> Result<Arc<dyn WritableStoreTrait>, StoreError> {
         let site = self.site(id)?;
-        Ok(Arc::new(WritableAgent::new(self.clone(), logger, site)?))
+        Ok(Arc::new(WritableStore::new(
+            self.clone(),
+            logger,
+            site,
+            self.registry.clone(),
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/subgraph_store.rs:1095]: Changed

File: store/postgres/src/writable.rs [hunk 3]
```diff
@@ -325,37 +351,471 @@ impl WritableStore {
...
+
+/// A queue that asynchronously writes requests queued with `push` to the
+/// underlying store and allows retrieving information that is a combination
+/// of queued changes and changes already committed to the store.
+struct Queue {
+    store: Arc<SyncStore>,
+    /// A queue of pending requests. New requests are appended at the
+    /// back, and popped off the front for processing. That means that
+    /// going front-to-back, block numbers in the requests are
+    /// increasing.
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/writable.rs:426]: That comment was a leftover from a previous incarnation where reverts didn't go into the queue. Fixed the comment

File: store/postgres/src/writable.rs [hunk 4]
```diff
@@ -325,37 +351,471 @@ impl WritableStore {
...
+                    ),
+                    Request::Revert {
+                        block_ptr,
+                        firehose_cursor,
+                    } => queue
+                        .store
+                        .revert_block_operations(block_ptr.clone(), firehose_cursor.as_deref()),
+                };
+                // Remove the tombstone that take_front left in place
+                queue.queue.pop().await;
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/writable.rs:472]: Yes, that is indeed a bit tricky. I just added this comment to the loop:
```rust
                // We peek at the front of the queue, rather than pop it
                // right away, so that query methods like `get` have access
                // to the data while it is being written. If we popped here,
                // these methods would not be able to see that data until
                // the write transaction commits, causing them to return
                // incorrect results.
```

File: store/postgres/src/writable.rs [hunk 5]
```diff
@@ -325,37 +351,471 @@ impl WritableStore {
...
+            poisoned: AtomicBool::new(false),
+        };
+        let queue = Arc::new(queue);
+
+        // Use a separate instance of the `StopwatchMetrics` for background
+        // work since that has its own call hierarchy, and using the
+        // foreground metrics will lead to incorrect nesting of sections
+        let stopwatch =
+            StopwatchMetrics::new(logger, queue.store.site.deployment.clone(), registry);
+        graph::spawn(start_writer(queue.clone(), stopwatch));
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/writable.rs:497]: Changed

File: graph/src/util/bounded_queue.rs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        permit.forget();
+        self.pop_semaphore.add_permits(1);
+    }
+
+    pub async fn wait_empty(&self) -> Result<(), ()> {
+        self.push_semaphore
+            .acquire_many(self.capacity as u32)
+            .await
+            .map(|_| ())
+            .map_err(|_| ())
```
[the remark this answers is not in the document]
Comment (lutter) [reply graph/src/util/bounded_queue.rs:85]: true - changed that and the method signature

File: store/postgres/src/writable.rs [hunk 7]
```diff
@@ -325,37 +351,471 @@ impl WritableStore {
...
+        Ok(())
+    }
+
+    /// Wait for the background writer to finish processing queued entries
+    async fn wait(&self) -> Result<(), StoreError> {
+        self.queue
+            .wait_empty()
+            .await
+            .map_err(|()| StoreError::Canceled)?;
+        match self.write_err.lock().unwrap().take() {
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/writable.rs:515]: heh .. yeah, why not?

File: store/postgres/src/writable.rs [hunk 8]
```diff
@@ -325,37 +351,471 @@ impl WritableStore {
...
+        self.check_err()?;
+        self.queue.push(Arc::new(req)).await;
+        Ok(())
+    }
+
+    /// Wait for the background writer to finish processing queued entries
+    async fn wait(&self) -> Result<(), StoreError> {
+        self.queue
+            .wait_empty()
+            .await
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/writable.rs:513]: I did just that: clear the queue on error

File: store/postgres/src/writable.rs [hunk 9]
```diff
@@ -325,37 +351,471 @@ impl WritableStore {
...
+            write_err,
+            poisoned: AtomicBool::new(false),
+        };
+        let queue = Arc::new(queue);
+
+        // Use a separate instance of the `StopwatchMetrics` for background
+        // work since that has its own call hierarchy, and using the
+        // foreground metrics will lead to incorrect nesting of sections
+        let stopwatch =
+            StopwatchMetrics::new(logger, queue.store.site.deployment.clone(), registry);
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/writable.rs]: I made it so that you can now add a 'stage' to StopwatchMetrics. LMK what you think

**Comment by lutter** [comment]
Addressed all review comments

File: store/postgres/src/writable.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (leoyvens) [reply store/postgres/src/writable.rs]: A label is very smart!

File: store/postgres/src/writable.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (leoyvens) [reply store/postgres/src/writable.rs:479]: Thanks, that really is subtle.

> // Remove the tombstone that take_front left in place

This comment is still cryptic to me.

File: graph/src/util/bounded_queue.rs [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+            let pushed = queue.len();
+            queue.clear();
+            pushed
+        };
+        self.push_semaphore.add_permits(pushed);
+        let _permits = self
+            .pop_semaphore
+            .acquire_many(pushed as u32)
+            .await
+            .expect("we never close the pop_semaphore");
```
Comment (leoyvens) [inline graph/src/util/bounded_queue.rs:140]: Why acquire these permits only to drop them?

File: store/postgres/src/writable.rs [hunk 11]
```diff
@@ -325,37 +351,477 @@ impl WritableStore {
...
+                        stopwatch.cheap_clone(),
+                        data_sources,
+                        deterministic_errors,
+                    ),
+                    Request::Revert {
+                        block_ptr,
+                        firehose_cursor,
+                    } => queue
+                        .store
+                        .revert_block_operations(block_ptr.clone(), firehose_cursor.as_deref()),
```
Comment (leoyvens) [inline store/postgres/src/writable.rs:476]: Now that this is called from a non-blocking tokio task, `transact_block_operations` and  `revert_block_operations` should be asyncified.

File: store/postgres/src/writable.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/writable.rs:479]: I changed it to
> // The request has been handled. It's now safe to remove it
> // from the queue

File: graph/src/util/bounded_queue.rs (hunk 10 above)
Comment (lutter) [reply graph/src/util/bounded_queue.rs:140]: I forgot to forget them. Fixed

File: store/postgres/src/writable.rs (hunk 11 above)
Comment (lutter) [reply store/postgres/src/writable.rs:476]: I tried asyncifying `transact_block_operations` for a bit, but it requires that `DeploymentStore.transact_block_operations` clone all its arguments to own them because they get put into a background task since Rust doesn't understand that the `with_conn(|_, _| ...).await` that uses the arguments has to return before `transact_block_operations` returns and wants references with a static lifetime.

Rather than copy large amounts of data, I think we should hold off on asyncifying.

Comment (leoyvens) [reply store/postgres/src/writable.rs:476]: Then we should spawn the writer in a blocking task.

File: graph/src/util/bounded_queue.rs [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    /// therefore not do any slow work.
+    pub fn fold<B, F>(&self, init: B, f: F) -> B
+    where
+        F: FnMut(B, &T) -> B,
+    {
+        let queue = self.queue.lock().unwrap();
+        queue.iter().rev().fold(init, f)
+    }
+
+    pub async fn clear(&self) {
```
Comment (leoyvens) [inline graph/src/util/bounded_queue.rs:128]: The way we use it wouldn't manifest this, but in principle calling `clear` and `pop` concurrently could cause pop to panic, since `clear` will empty the queue without first acquiring the permits. If it tried to acquire the permits first we'd have a different potential bug, where the permit count may change between it being checked and the permits actually being acquired, and then `clear` could deadlock.

I'm thinking that `BoundedQueue` shouldn't intrinsically support `clear`, but rather the writer should call `pop` in a loop until the queue is empty when clearing it. This way we can justify the assumption that neither `pop` nor `push` are being called concurrently to this.

Comment (lutter) [reply graph/src/util/bounded_queue.rs:128]: I added a `try_pop` (non-blocking `pop`) and made `clear` a wrapper around calling that in a loop; since there's no `await` on that code path now, there shouldn't be any possibility for races.

File: store/postgres/src/writable.rs (hunk 11 above)
Comment (lutter) [reply store/postgres/src/writable.rs:476]: I didn't put the whole writer into a blocking task since that blocks the process from exiting; rather, I run the actual store operations as a blocking task now.

**Comment by lutter** [comment]
Addressed all comments; I'll rebase after the review to make it not too hard to follow the convo.

File: store/postgres/src/writable.rs [hunk 13]
```diff
@@ -453,37 +481,27 @@ impl Queue {
...
-                        firehose_cursor,
-                    } => queue
-                        .store
-                        .revert_block_operations(block_ptr.clone(), firehose_cursor.as_deref()),
-                };
+                let req = queue.queue.peek().await;
+                let store = queue.store.cheap_clone();
+                let stopwatch = stopwatch.cheap_clone();
+                let res =
+                    graph::spawn_blocking(
```
Comment (leoyvens) [inline store/postgres/src/writable.rs:488]: The naming of the `graph::task_spawn` methods is not very consistent, but you should use `graph::spawn_blocking_allow_panic` here which takes a closure rather than this which takes a future.

Or if you'd rather abort on panic you'd need to add a new helper that takes an `FnOnce` and does that, though catching the panic and propagating as an error seems like what we'd want.

**Review by leoyvens — CHANGES_REQUESTED** [review]

File: store/postgres/src/writable.rs (hunk 13 above)
Comment (lutter) [reply store/postgres/src/writable.rs:488]: I didn't realize that - I just updated the code; now, panics are turned into `StoreError` which seems like a much better way to handle them anyway

**Review by leoyvens — APPROVED** [review]

**Comment by lutter** [comment]
Rebased to latest master

**Comment by lutter** [comment]
@leoyvens I made some changes to this PR since you reviewed it. Could you have a look at the last 4 commits? I think the other ones didn't change, they were just rebased.

**Review by leoyvens — COMMENTED** [review]
Seems like it would be helpful to rename `Request::Revert` to `Request::RevertTo`.

File: store/postgres/src/writable.rs [hunk 14]
```diff
@@ -390,17 +401,13 @@ impl BlockTracker {
...
-            self.block - 1
-        }
+        self.block
     }
 
     /// Return `true` if a write at this block will be visible, i.e., not
     /// reverted by a previous queue entry
     fn visible(&self, block_ptr: &BlockPtr) -> bool {
-        self.revert > block_ptr.number
+        self.revert >= block_ptr.number
```
[the remark this answers is not in the document]
Comment (lutter) [reply store/postgres/src/writable.rs:410]: Hah .. Yoda order :) Yes, reversing the comparison looks better

**Comment by lutter** [comment]
Added a commit with the two suggestions and rebased to latest master

**Review by leoyvens — APPROVED** [review]
